### DISCUSSION TITLE
Confused about `config =` on new plugins

### DISCUSSION THREAD

**Discussion by florentc** [wbthomason/packer.nvim#506] [Q&A] [answered]
Hello,

I am confused about how to use packer in the following case:

Starting from an empty configuration:

```lua
require('packer').startup(function()
    use 'wbthomason/packer.nvim'
end
```

I want to use a new plugin `theplugin` and configure it. I add something like:

```lua
require('packer').startup(function()
    use 'wbthomason/packer.nvim'
    use {
        'theauthor/theplugin',
        config = require('theplugin').setup { ... }
    }
end
```

I need to source the modified file or restart nvim before I can run `:PackerCompile` or `:PackerSync`. Since the plugin has yet to be installed, I get a `module 'theplugin' not found`. Running `:PackerCompile` and/or `:PackerSync` does not work then.

I have to remove `config = require('theplugin').setup { ... }`, compile/sync to install the plugin, and then I can put the config line back (and this time obviously the module is found because the plugin has been installed). Just like with traditional vim plugin managers.

Therefore, I don't get the point of the `config` key. I thought the `config` function would only be called if the plugin is installed and its modules ready for use.

Am I missing something?

Thank you for your help.

**Comment by wbthomason** [comment, marked as answer]
This is a reasonable question - partially, the flow you used is a bit different than the usual, and partially, `config` doesn't do quite what you think (though perhaps this should change).

Usually, for a new plugin install, I would
1. Write the plugin spec
2. Load the plugin specification file (restarting Neovim if I have to because of Lua's module caching)
3. Run `PackerSync` to first clone the plugin and then generate (and source) the compiled file

In this case, everything goes "as expected". I'm guessing that you probably ran `PackerCompile` first, which generated an invalid compiled file because the plugin wasn't installed? (that said, the `try_loadstring` function we use in `packer_compiled` should have prevented a failing config from affecting others, so I'm not sure why you were unable to run `packer` commands, as you say).

The other bit of this is that, right now, `config` etc. does not check if a plugin is installed before trying to run. This might be worth adding; let me throw together a quick PR with an idea I have and see if it works as you'd hope.

**Reply by wbthomason** [reply]
Ok, check out #508 which makes `config` behave more like you (reasonably) expect!

**Reply by florentc** [reply]
Thank you very much for your quick reply and implementing a new dedicated feature that fast!

I switched to the feature branch and did the following:

1. Add a new plugin to the plugin spec 

```lua
use {
    'theauthor/theplugin',
    config = require('theplugin').setup { ... }
}
```

2. Save the file
3. Restart neovim

I get the "module not found error" at startup as before. Before I can run `PackerSync` and `PackerCompile`. If I run `PackerSync` it offers to remove the plugins specified after the error line.

I guess the new feature prevents issues when calling compile then sync (as I used to do) but it seems I am also doing something wrong before compilation is involved.

**Reply by wbthomason** [reply]
Ah, I see - this is a slightly different problem. Here, the issue is just normal Lua evaluation - when you run the plugin spec, Lua eagerly evaluates the `config` key's value to get the config function (note: this is not the same as running the config function). I should've noticed and commented on this in your original post, but your `config` key is slightly wrong - `config` (and `setup`) must be either **strings** or **functions**, whereas your spec eagerly evaluates a function call to `setup`. You probably want `config = function() require('theplugin').setup {} end` instead.

**Reply by florentc** [reply]
Oh OK I see! I am so sorry to have bothered you and wasted time about such a basic mistake. In the documentation of packer and other plugins, such calls are always encapsulated inside a function. My brain has repeatedly ignored that for some reason. Thank you again for your help and your work. The new feature does come in handy.

**Reply by wbthomason** [reply]
No, no worries! `packer` has a definite learning curve; I'm happy to help. I'm glad things are working now.
